README.md
<p>
	<a href="https://travis-ci.org/mageddo/dns-proxy-server"><img src="https://travis-ci.org/mageddo/dns-proxy-server.svg?branch=master" alt="Build Status" /></a>
</p>

### Features

DPS is a end user(developers, Server Administrators) DNS server tool to develop systems with docker solving
docker containers hostnames:

* Solve hostnames from local configuration database
* Solve hostnames from docker containers using docker **hostname** option or **HOSTNAMES** env
* Solve hostnames from a list of configured DNS servers(as a proxy) if no answer of two above
* [Solve hostnames using wildcards](http://mageddo.github.io/dns-proxy-server/docs/features#solve-hostnames-using-wildcards)
* [Graphic interface to manage it](http:/127.0.0.1:5380/static/)
	* List and edit DNS local entries
* [Solve host machine IP using `host.docker` hostname](http://mageddo.github.io/dns-proxy-server/docs/features#solve-host-machine-ip-from-anywhere)

**For more details see** [the Documentation ](http://mageddo.github.io/dns-proxy-server/docs/features) or [Release Notes](RELEASE-NOTES.md) 

![](https://i.imgur.com/aR9dl0O.png)

### Running it

```bash
$ docker run --rm --hostname dns.mageddo --name dns-proxy-server -p 5380:5380 \
-v /var/run/docker.sock:/var/run/docker.sock \
-v /etc/resolv.conf:/etc/resolv.conf \
defreitas/dns-proxy-server
```

then try it out

```bash
$ ping dns.mageddo
PING dns.mageddo (172.17.0.4) 56(84) bytes of data.
64 bytes from 172.17.0.4: icmp_seq=1 ttl=64 time=0.063 ms
64 bytes from 172.17.0.4: icmp_seq=2 ttl=64 time=0.074 ms
64 bytes from 172.17.0.4: icmp_seq=3 ttl=64 time=0.064 ms
```

[Checkout the full running it documentation for more details](http://mageddo.github.io/dns-proxy-server/docs/running.html)

### Requirements
* Linux/Windows
* Docker 1.9.x (Only if you run DPS using docker or want to solve docker containers hostname using DPS)

### DNS resolution order
DNS  Proxy Server follow the below order to solve hostnames

* Try to solve the hostname from **docker** containers
* Then from local database file
* Then from 3rd configured remote DNS servers

### Documents
* [Latest Rest API Features](http://mageddo.github.io/dns-proxy-server/docs/api/)
* [Coding](docs/developing) at DNS Proxy Server

### MAC Support
Based on users feedback, DPS don't work on MAC, unfortunatly I don't have a MAC computer to work on that, 
if you want to contribute please try to fix it then open a pull request, sorry for the inconvenience.


RELEASE-NOTES.md
### 2.18.6
* Fixing gateway IP resolution when not in DPS network (#186)

### 2.18.5
* Fixing unnecessary stacktraces were being logged
* Answering NXDOMAIN when no answers were found
* Fixing logging file trace

### 2.18.4
* Bumping github-cli to fix releasing

### 2.18.3
* Resolving docker services using configured DPS domain
* Fix presence check of config setting "domain"
* Adding working around coordinates for resolv.conf at the docs
* Fixing releasing

### 2.18.2
* Fixing wrong mapping on `logLevel` property

### 2.18.1
* Change log level before try to log something

### 2.18.0
* Feature: Multiple environments, now you can setup a group of hostnames and save it to a environment, then you can 
create a new environment and switch between them, very useful when working on different contexts switching from QA to PROD,
for example, [see the docs](http://mageddo.github.io/dns-proxy-server/2.18/en/2-features/multiple-environments/)

### 2.17.4
* Clearing cache for resolvers when the config file is saved

### 2.17.3
* Separating the build image from final image, removing unnecessary bash command

### 2.17.2
* Fixing docker build was using deprecated apt-get option

### 2.17.1
* Reducing docker image size by 20%~

### 2.17.0
* Go version upgrade from 1.11 to 1.12

### 2.16.0
* Upgrading docker images to debian-10-slim
* Reducing up to 30% on image size

### 2.15.0
* Decreasing chance of acl issues by giving priority to answer ip of bridge networks over overlay ones
* Now DPS can have your own network this way it can access and be accessed
by all docker containers, **not** enabled by default [see the docs](http://mageddo.github.io/dns-proxy-server/2.15/en/2-features/dps-network-resolution/) 

### 2.14.6
* Fixing ping slowness

### 2.14.5
* Fixing docker hub push

### 2.14.4
* Fixing log level wasn't being respected 

### 2.14.2
* Ability to specify remote server port
* Introducing storage api v2
* Refactoring the docs to use Hugo templates 
 
### 2.14.1
* Fixing nil pointer when remote server get timeout (#126)
* Simplify bug report
* Fixing nil pointer when remote server returns timeout

### 2.14.0
* Making some refactoring facilitating to the feature requested at #121
* Fixing nil pointer sometimes when the hostname were not found

### 2.13.2
* Fixing broken answer when hostname is not found
* Fixing ping slowness
 
### 2.13.1
* Make sure value column will not break the table (#116)

### 2.13.0
* Support for CNAME on local entries, [see the docs](https://github.com/mageddo/dns-proxy-server/blob/7dacc2c/docs/features.md#manager-customer-dns-records)

### 2.12.0
* Possibility to change container hostname domain, [see the docs](https://github.com/mageddo/dns-proxy-server/blob/70a0ff8/docs/features.md#access-container-by-its-container-name--service-name)

### 2.11.0
* Now you can customize host machine hostname, see [the docs](https://github.com/mageddo/dns-proxy-server/blob/fa1e044b/docs/features.md#solve-host-machine-ip-from-anywhere)
* Increased default loglevel to INFO

### 2.10.3
* Build arm images on travis cause docker hub haven't support

### 2.10.2
* Fixing binaries were generated for wrong arch

### 2.10.1
* Official support for ARM

### 2.9.1
* Supporting Multilevel wildcard
* Fixing ping slowness, bug introduced on **2.9.0**

### 2.9.0
* Now remote resolved names are cached respecting TTL
* Refactored local storage cache

### 2.8.0
* If your container have multiple networks you can specify which network to use when solving IP by specifying `dps.network` label

### 2.7.0
* Now you can access your container by it's container / docker-compose service name, syntax is `<container-name>.docker`

### 2.6.1
* Updating docs

### 2.6.0
* Now you can solve host machine IP from anywhere using host `host.docker`

### 2.5.4
* Organize some logs and auto reconfigure as default dns if resolvconf changes

### 2.5.3
* Fixing wildcard resolution were not solving main domain to local configuration, just the subdomains

### 2.5.2
* Fixing log level that stopped of work after **2.5.0**
* Fixing and increasing docs development instructions
* Fixing wildcard resolution were not solving main domain to docker container, just the subdomains

### 2.5.1
* Fixing ping slowness, takes more than 10 seconds to respond 

### 2.5.0
* Migrate to static logging tool

### 2.4.1
* Service restart command was with typo

### 2.4.0
* Enable/Disable log/set log path using `MG_LOG_FILE` env or `--log-file` command line option or json config
* Change log level using `MG_LOG_LEVEL` env or `--log-level` command line option or json config

### 2.3.3
* Domains wildcard support
If you register a hostname with `.` at start, then all subdomains will solve to that container/local storage entry

### 2.2.3
* Some times container hostname don't get registered at machine startup

### 2.2.2
* Cache Rest API v1 is exposed

### 2.2.1
* Preventing nil pointer when container inspection fails

### 2.2.0
* Increased code coverage
* Implementing cache at local hostnames and remote server resolution
* Considering TTL to invalidate hostname cache for local resolution

### 2.1.7
* All build and release process is made inside docker (no travis dependency)

### 2.1.6
* Refactor project structure to save dependencies in vendor folder

### 2.1.5
* Automating build with Travis

### 2.1.1
* Fix - `Error response from daemon: No such container...` message. see #29  
* Fix - hostname don't get removed when the container has killed. see #26  

### 2.1.0
* Turn publish port optional when running as service using docker mode

### 2.0.21
* BugFix - Service stopped of work in normal mode

### 2.0.20 
* Support for --version option that shows the current version
* Docker Compose is not required anymore to run DNS Proxy Server as a docker service

### 2.0.19
* Ability to customize remote server
* Fixing DNS solution order from (local, docker, remote) to (docker, local, remote)
* Now, at least docker 1.9 API v1.21 is necessary

### 2.0.18

utils/utils.go
package utils

import (
	"path/filepath"
	"os"
	"encoding/json"
	"io"
	"time"
	"github.com/mageddo/dns-proxy-server/utils/env"
	"strings"
	"os/signal"
	"syscall"
	"bytes"
	"github.com/mageddo/dns-proxy-server/flags"
)

var QTypeCodes = map[uint16] string {
0 : "TypeNone",
1 : "TypeA",
2 : "TypeNS",
3 : "TypeMD",
4 : "TypeMF",
5 : "TypeCNAME",
6 : "TypeSOA",
7 : "TypeMB",
8 : "TypeMG",
9 : "TypeMR",
10 : "TypeNULL",
12 : "TypePTR",
13 : "TypeHINFO",
14 : "TypeMINFO",
15 : "TypeMX",
16 : "TypeTXT",
17 : "TypeRP",
18 : "TypeAFSDB",
19 : "TypeX25",
20 : "TypeISDN",
21 : "TypeRT",
23 : "TypeNSAPPTR",
24 : "TypeSIG",
25 : "TypeKEY",
26 : "TypePX",
27 : "TypeGPOS",
28 : "TypeAAAA",
29 : "TypeLOC",
30 : "TypeNXT",
31 : "TypeEID",
32 : "TypeNIMLOC",
33 : "TypeSRV",
34 : "TypeATMA",
35 : "TypeNAPTR",
36 : "TypeKX",
37 : "TypeCERT",
39 : "TypeDNAME",
41 : "TypeOPT", // EDNS
43 : "TypeDS",
44 : "TypeSSHFP",
46 : "TypeRRSIG",
47 : "TypeNSEC",
48 : "TypeDNSKEY",
49 : "TypeDHCID",
50 : "TypeNSEC3",
51 : "TypeNSEC3PARAM",
52 : "TypeTLSA",
53 : "TypeSMIMEA",
55 : "TypeHIP",
56 : "TypeNINFO",
57 : "TypeRKEY",
58 : "TypeTALINK",
59 : "TypeCDS",
60 : "TypeCDNSKEY",
61 : "TypeOPENPGPKEY",
99 : "TypeSPF",
100 : "TypeUINFO",
101 : "TypeUID",
102 : "TypeGID",
103 : "TypeUNSPEC",
104 : "TypeNID",
105 : "TypeL32",
106 : "TypeL64",
107 : "TypeLP",
108 : "TypeEUI48",
109 : "TypeEUI64",
256 : "TypeURI",
257 : "TypeCAA",

249 : "TypeTKEY",
250 : "TypeTSIG",

// valid Question.Qtype only,
251 : "TypeIXFR",
252 : "TypeAXFR",
253 : "TypeMAILB",
254 : "TypeMAILA",
255 : "TypeANY",

32768 : "TypeTA",
32769 : "TypeDLV",
65535 : "TypeReserved",

}

var QClassCodes = map[uint16] string {

	// valid Question.Qclass,
	1 : "ClassINET",
	2 : "ClassCSNET",
	3 : "ClassCHAOS",
	4 : "ClassHESIOD",
	254 : "ClassNONE",
	255 : "ClassANY",
}

var RCodes  = map[uint16] string {
	// Message Response Codes.,
	0 : "RcodeSuccess",
	1 : "RcodeFormatError",
	2 : "RcodeServerFailure",
	3 : "RcodeNameError",
	4 : "RcodeNotImplemented",
	5 : "RcodeRefused",
	6 : "RcodeYXDomain",
	7 : "RcodeYXRrset",
	8 : "RcodeNXRrset",
	9 : "RcodeNotAuth",
	10 : "RcodeNotZone",
	16 : "RcodeBadSig", // TSIG
	//16 : "RcodeBadVers", // EDNS0
	17 : "RcodeBadKey",
	18 : "RcodeBadTime",
	19 : "RcodeBadMode", // TKEY
	20 : "RcodeBadName",
	21 : "RcodeBadAlg",
	22 : "RcodeBadTrunc", // TSIG
	23 : "RcodeBadCookie", // DNS Cookies
}

var opCodes  = map[uint16] string {
	// Message Opcodes. There is no 3.,
	0 : "OpcodeQuery",
	1 : "OpcodeIQuery",
	2 : "OpcodeStatus",
	4 : "OpcodeNotify",
	5 : "OpcodeUpdate",
}

var Sig = make(chan os.Signal)

func init(){
	signal.Notify(Sig, syscall.SIGINT, syscall.SIGTERM)
}

func DnsQTypeCodeToName(code uint16) string {
	return QTypeCodes[code]
}

func GetCurrentPath() string {

	currDIr := os.Getenv(env.MG_WORK_DIR)
	if len(currDIr) != 0 {
		return currDIr
	}
	currentPath, _ := filepath.Abs(filepath.Dir(os.Args[0]))
	return currentPath

}

func GetPath(path string) string {
	if !strings.HasPrefix(path, "/") {
		path = "/" + path
	}
	currentPath := GetCurrentPath();
	if strings.HasSuffix(currentPath, "/") {
		currentPath = currentPath[0:len(currentPath)-1]
	}
	return currentPath + path
}

func GetJsonEncoder(w io.Writer) *json.Encoder {
	enconder := json.NewEncoder(w)
	if !flags.IsTestVersion() {
		enconder.SetIndent("", "\t")
	}
	return enconder
}

func GetUUID() int64 {
	return time.Now().UnixNano()
}

func Copy(src, dst string) error {
	in, err := os.Open(src)
	if err != nil { return err }
	defer in.Close()
	out, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
	if err != nil { return err }
	defer out.Close()
	_, err = io.Copy(out, in)
	cerr := out.Close()
	if err != nil { return err }
	return cerr
}

func CreateExecutableFile(sourceData, dst string) error {
	out, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
	if err != nil { return err }
	defer out.Close()
	_, err = io.Copy(out, bytes.NewReader([]byte(sourceData)))
	cerr := out.Close()
	if err != nil { return err }
	return cerr
}

func WriteToFile(sourceData, dst string) error {
	out, err := os.OpenFile(dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0766)
	if err != nil { return err }
	defer out.Close()
	_, err = io.Copy(out, bytes.NewReader([]byte(sourceData)))
	cerr := out.Close()
	if err != nil { return err }
	return cerr
}


// Difference between b and a in milliseconds
//
// b must be greater than a
//
func DiffMillis(a, b time.Time) int64 {
	na, nb := a.UnixNano(), b.UnixNano()
	return (nb - na) / int64(time.Nanosecond * time.Millisecond)
}
